lib/db_connection/connection.ex

```diff
@@ -351,6 +353,17 @@ defmodule DBConnection.Connection do
     handle_timeout(s)
   end
 
+  @doc false
+  @impl :gen_statem
+  # If client is :closed then the connection was previouly disconnected
+  # and cleanup is not required.
+  def terminate(_, _, %{client: :closed}), do: :ok
+
+  def terminate(reason, _, s) do
+    %{mod: mod, state: state} = s
```

Comment:
Should we call `notify_connection_listeners(:disconnected, s)`? Thoughts @whatyouhide?

Reply:
Maybe not, since it depends on the parent trapping exits, the behaviour may not be consistent. It is better if people tracking listeners also monitors the process.